mozilla_django_oidc/views.py

```diff
@@ -68,18 +64,28 @@ def get(self, request):
                 auth.logout(request)
             assert not request.user.is_authenticated
         elif 'code' in request.GET and 'state' in request.GET:
-            kwargs = {
-                'request': request,
-                'nonce': nonce,
-            }
 
-            if 'oidc_state' not in request.session:
+            # Check instead of "oidc_state" check if the "oidc_states" session key exists!
+            if 'oidc_states' not in request.session:
                 return self.login_failure()
 
-            if request.GET['state'] != request.session['oidc_state']:
-                msg = 'Session `oidc_state` does not match the OIDC callback state'
+            # State and Nonce are stored in the session "oidc_states" dictionary.
+            # State is the key, the value is a dictionary with the Nonce in the "nonce" field.
+            state = request.GET.get('state')
+            if state not in request.session['oidc_states']:
+                msg = 'OIDC callback state not found in session `oidc_states`!'
                 raise SuspiciousOperation(msg)
 
+            # Get the nonce from the dictionary for further processing and delete the entry to
+            # prevent replay attacks.
+            nonce = request.session['oidc_states'][state]['nonce']
+            del request.session['oidc_states'][state]
```

Comment:
Is it possible to move this `del` much higher up in the method?  In particular, before you you test if `error` or `code` are in the query parameters.

If the purpose of a nonce is to prevent replay attacks, then it feels safer to remove the nonce from the `oidc_states` as early as you possibly can.  Right now I cannot figure out if an attacker could take advantage of your implementation (probably not!).  But if you can move it further up then you might prevent bugs or replay attack vulnerabilities from creeping into the code in future.

Reply:
I don't think we can move it further up. All the code above does, is to retrieve the state/nonce combination from the dictionary.

I'm actually not really sure how much the nonce helps here anyways.